mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,677 @@
+    Non4KL2Entry,
+    /// The caller specified an invalid physical address (too large for example)
+    InvalidAddress,
+    /// The platform initialization hasn't been completed. The implementation is expected
+    /// to initialize pointers to the MTT L2 and L1 page pool using platform-FW tables.
+    PlatformInitNotCompleted,
+    /// The platform didn't allocate sufficient L1 page pool pages to map the specified address
+    InsufficientL1Pages,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Size of the larger memory range containing the page.
+/// This is returned following a successful call to set the type of the page
+/// and can be used for bookkeeping purposes. This can be useful for performance
+/// optimizations like combining or eliding ECALLs to invalidate the MTT.
+pub enum MemoryRangeType {
+    /// The page is part of an 1G non-confidential range
+    NonConfidential1G,
+    /// The page is part of an 1G confidential range
+    Confidential1G,
+    /// The page is part of a 64MB non-confidential range
+    NonConfidential64MB,
+    /// The page is part of a 64MB confidential range
+    Confidential64MB,
+    /// The 4K page is non-confidential
+    NonConfidential4K,
+    /// The 4K page is confidential
+    Confidential4K,
+}
+
+const MTT_L2_ENTRY_COUNT: usize = 1024 * 1024;
+const MTT_L1_ENTRY_COUNT: usize = 512;
+// Struct to represent a L1 page
+// Each entry represents 32 4K pages, which represents a total of 64MB of address space.
+// Each page uses a 2-bit encoding, with '00b as non-confidential and '01b as confidential ('11b is illegal).
+struct MttL1Page {
+    base: [u64; MTT_L1_ENTRY_COUNT],
+}
+struct MttCoreInner {
+    mtt_l2_base: &'static mut [u64],
+    mtt_l1_page_pool_base: &'static mut [MttL1Page],
+}
+
+struct MttCore {
+    inner: Mutex<MttCoreInner>,
+}
+
+static mut MTT_CORE: Once<MttCore> = Once::new();
+// Utility structure for for L2 manipulation
+struct MttHelper<'a> {
+    // Type of MTT L2 entry for the physical address in `phys_addr`
+    l2_entry: MttL2EntryType,
+    phys_addr: SupervisorPageAddr,
+    inner: MutexGuard<'a, MttCoreInner>,
+}
+
+#[allow(dead_code)]
+impl MttCore {
+    /// Creates a new MttL1PagePool object.
+    /// # Safety -- the caller must ensure that:
+    /// 1: `l2_base` is a physically contiguous and reserved 8MB region allocated by platform-FW corresponding to the MTT L2.
+    /// 2: `l1_pool_base` is a valid pointer to a contiguous reserved region of `l1_pool_page_count` 4K-pages. Note that for computation purposes,
+    /// we assume that the platform memory is a contiguous range beginning at physical address 0 and ending at the top of physical memory. The caller must
+    /// allocate 16 4K pages in the pool for each 1GB of addressable space.
+    /// 4: `l2_base_addr` and `l1_pool_base_addr` must not be aliased elsewhere, and must remain allocated until a platform reset.
```

Comment:
This is a rather complicated safety requirement. Could the new consume slices and have the slice creation of each slice be a separate unsafe block in the caller? Or maybe this can consume `Page`s?

Reply:
Updated per offline conversation.